src/combinators/mod.rs

```diff
@@ -0,0 +1,135 @@
+
+    Eventual::spawn(|mut writer| async move {
+        while let Ok(v) = source.next().await {
+            writer.write(f(v).await);
+        }
+    })
+}
+
+pub fn timer(interval: Duration) -> Eventual<Instant> {
+    Eventual::spawn(move |mut writer| async move {
+        loop {
+            writer.write(Instant::now());
+            time::sleep(interval).await;
+        }
+    })
+}
+
+// TODO: Put this in a macro to de-duplicate A/B and support more arguments.
+pub fn join<A, Ar, B, Br>(a: Ar, b: Br) -> Eventual<(A, B)>
+where
+    A: Value,
+    B: Value,
+    Ar: IntoReader<Output = A>,
+    Br: IntoReader<Output = B>,
+{
+    let mut a = a.into_reader();
+    let mut b = b.into_reader();
+
+    Eventual::spawn(move |mut writer| async move {
+        let mut count = 0;
+        let mut ab = (None, None);
+
+        let mut ab = loop {
+            select! {
+                a_value = a.next() => {
+                    match a_value {
+                        Ok(a_value) => {
+                            if ab.0.replace(a_value).is_none() {
+                                count += 1;
+                            }
+                        },
+                        Err(_) => { return ; }
+                    }
+                }
+                b_value = b.next() => {
+                    match b_value {
+                        Ok(b_value) => {
+                            if ab.1.replace(b_value).is_none() {
+                                count += 1;
+                            }
+                        },
+                        Err(_) => {
+                            return;
+                        }
+                    }
+                }
+            }
+            if count == 2 {
+                break (ab.0.unwrap(), ab.1.unwrap());
+            }
+        };
+        loop {
+            writer.write(ab.clone());
+
+            select! {
+                a_value = a.next() => {
+                    match a_value {
+                        Ok(a_value) => { ab.0 = a_value; }
+                        Err(_) => { return; }
+                    }
+                }
+                b_value = b.next() => {
+                    match b_value {
+                        Ok(b_value) => { ab.1 = b_value; }
+                        Err(_) => { return; }
+                    }
+                }
+            }
+        }
+    })
+}
+
+pub fn throttle<E>(read: E, duration: Duration) -> Eventual<E::Output>
+where
+    E: IntoReader,
+{
+    let mut read = read.into_reader();
+
+    Eventual::spawn(move |mut writer| async move {
+        while let Ok(mut next) = read.next().await {
+            let end = time::Instant::now() + duration;
+            loop {
+                // Allow replacing the value until the time is up. This
+                // necessarily introduces latency but de-duplicates when there
+                // are intermittent bursts. Not sure what is better. Matching
+                // common-ts for now.
```

Comment:
Yep. The version that doesn't add latency is simpler to write (it looks exactly like timer but spits out next().await instead of the instant). For UIs and similar I think this pattern works better because it doesn't pick up the smallest first change. A better API that is the best of both worlds might have a min/max latency that will once a change is available spit out a value if there has been no change for min latency while never waiting longer than max latency. Not urgent in any case.